From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Baocheng Su <baocheng.su@siemens.com>
Date: Thu, 9 Jul 2026 16:33:52 +0800
Subject: [PATCH] iot2050: add GPIO chardev v2 and revision-aware pinmux
 support

Add GPIO character device v2 support for IOT2050 GPIO and IO
expander handling.

Replace the fixed pinmux table with board-specific pin mappings,
support revision-dependent pinmux configuration, and configure GPIO
bias modes through the GPIO chardev v2 interface.

Signed-off-by: Baocheng Su <baocheng.su@siemens.com>
---
 src/arm/siemens/iot2050.c | 1862 ++++++++++++++-----------------------
 1 file changed, 686 insertions(+), 1176 deletions(-)

diff --git a/src/arm/siemens/iot2050.c b/src/arm/siemens/iot2050.c
index df60ed0e51e0..d471258cefa3 100644
--- a/src/arm/siemens/iot2050.c
+++ b/src/arm/siemens/iot2050.c
@@ -21,24 +21,27 @@
  * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
  * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
+#include <errno.h>
+#include <limits.h>
+#include <mraa/types.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/ioctl.h>
 #include <sys/mman.h>
-#include <limits.h>
-#include <mraa/types.h>
 
-#include "common.h"
-#include "gpio/gpio_chardev.h"
 #include "arm/siemens/iot2050.h"
 #include "arm/siemens/platform.h"
+#include "common.h"
+#include "gpio/gpio_chardev.h"
+#include "gpio/gpio_chardev_v2.h"
+// #include "mraa_internal_types.h"
 
-#define PINMUX_GROUP_MAIN       (0)
-#define PINMUX_GROUP_WAKUP      (1)
+#define PINMUX_GROUP_MAIN (0)
+#define PINMUX_GROUP_WAKUP (1)
 
 typedef struct {
     uint8_t     group;
-    uint16_t    index;
-    uint16_t    pinmap;
+    uint16_t index;
     int8_t      mode[MAX_MUX_REGISTER_MODE];
     const char *debugfs_path[MAX_MUX_REGISTER_MODE];
     const char *pmx_function[MAX_MUX_REGISTER_MODE];
@@ -46,7 +49,134 @@ typedef struct {
 }regmux_info_t;
 
 static void *pinmux_instance = NULL;
-static regmux_info_t pinmux_info[MRAA_IOT2050_PINCOUNT];
+static regmux_info_t* pinmux_info[MRAA_IOT2050_PINCOUNT] = {
+    [0] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                             17,
+                             { 7, 4, -1, -1, -1 },
+                             { "4301c000.pinctrl-pinctrl-single", "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d0-gpio", "d0-uart0-rxd", NULL, NULL, NULL },
+                             { "d0-gpio", "d0-uart0-rxd", NULL, NULL, NULL } },
+    [1] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                             18,
+                             { 7, 4, -1, -1, -1 },
+                             { "4301c000.pinctrl-pinctrl-single", "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d1-gpio", "d1-uart0-txd", NULL, NULL, NULL },
+                             { "d1-gpio", "d1-uart0-txd", NULL, NULL, NULL } },
+    [2] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                             19,
+                             { 7, 4, -1, -1, -1 },
+                             { "4301c000.pinctrl-pinctrl-single", "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d2-gpio", "d2-uart0-ctsn", NULL, NULL, NULL },
+                             { "d2-gpio", "d2-uart0-ctsn", NULL, NULL, NULL } },
+    [3] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                             21,
+                             { 7, 4, -1, -1, -1 },
+                             { "4301c000.pinctrl-pinctrl-single", "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d3-gpio", "d3-uart0-rtsn", NULL, NULL, NULL },
+                             { "d3-gpio", "d3-uart0-rtsn", NULL, NULL, NULL } },
+    [4] = &(regmux_info_t) { PINMUX_GROUP_MAIN,
+                             33,
+                             { 7, 5, -1, -1, -1 },
+                             { "11c000.pinctrl-pinctrl-single", "11c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d4-gpio", "d4-ehrpwm0-a", NULL, NULL, NULL },
+                             { "d4-gpio", "d4-ehrpwm0-a", NULL, NULL, NULL } },
+    [5] = &(regmux_info_t) { PINMUX_GROUP_MAIN,
+                             35,
+                             { 7, 5, -1, -1, -1 },
+                             { "11c000.pinctrl-pinctrl-single", "11c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d5-gpio", "d5-ehrpwm1-a", NULL, NULL, NULL },
+                             { "d5-gpio", "d5-ehrpwm1-a", NULL, NULL, NULL } },
+    [6] = &(regmux_info_t) { PINMUX_GROUP_MAIN,
+                             38,
+                             { 7, 5, -1, -1, -1 },
+                             { "11c000.pinctrl-pinctrl-single", "11c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d6-gpio", "d6-ehrpwm2-a", NULL, NULL, NULL },
+                             { "d6-gpio", "d6-ehrpwm2-a", NULL, NULL, NULL } },
+    [7] = &(regmux_info_t) { PINMUX_GROUP_MAIN,
+                             43,
+                             { 7, 5, -1, -1, -1 },
+                             { "11c000.pinctrl-pinctrl-single", "11c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d7-gpio", "d7-ehrpwm3-a", NULL, NULL, NULL },
+                             { "d7-gpio", "d7-ehrpwm3-a", NULL, NULL, NULL } },
+    [8] = &(regmux_info_t) { PINMUX_GROUP_MAIN,
+                             48,
+                             { 7, 5, -1, -1, -1 },
+                             { "11c000.pinctrl-pinctrl-single", "11c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d8-gpio", "d8-ehrpwm4-a", NULL, NULL, NULL },
+                             { "d8-gpio", "d8-ehrpwm4-a", NULL, NULL, NULL } },
+    [9] = &(regmux_info_t) { PINMUX_GROUP_MAIN,
+                             51,
+                             { 7, 5, -1, -1, -1 },
+                             { "11c000.pinctrl-pinctrl-single", "11c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                             { "d9-gpio", "d9-ehrpwm5-a", NULL, NULL, NULL },
+                             { "d9-gpio", "d9-ehrpwm5-a", NULL, NULL, NULL } },
+    [10] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              39,
+                              { 7, 0, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single",
+                                "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                              { "d10-gpio", "d10-spi0-cs0", NULL, NULL, NULL },
+                              { "d10-gpio", "d10-spi0-cs0", NULL, NULL, NULL } },
+    [11] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              37,
+                              { 7, 0, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single",
+                                "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                              { "d11-gpio", "d11-spi0-d0", NULL, NULL, NULL },
+                              { "d11-gpio", "d11-spi0-d0", NULL, NULL, NULL } },
+    [12] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              38,
+                              { 7, 0, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single",
+                                "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                              { "d12-gpio", "d12-spi0-d1", NULL, NULL, NULL },
+                              { "d12-gpio", "d12-spi0-d1", NULL, NULL, NULL } },
+    [13] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              36,
+                              { 7, 0, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single",
+                                "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL },
+                              { "d13-gpio", "d13-spi0-clk", NULL, NULL, NULL },
+                              { "d13-gpio", "d13-spi0-clk", NULL, NULL, NULL } },
+    [14] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              33,
+                              { 7, -1, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL, NULL },
+                              { "a0-gpio", NULL, NULL, NULL, NULL },
+                              { "a0-gpio", NULL, NULL, NULL, NULL } },
+    [15] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              32,
+                              { 7, -1, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL, NULL },
+                              { "a1-gpio", NULL, NULL, NULL, NULL },
+                              { "a1-gpio", NULL, NULL, NULL, NULL } },
+    [16] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              31,
+                              { 7, -1, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL, NULL },
+                              { "a2-gpio", NULL, NULL, NULL, NULL },
+                              { "a2-gpio", NULL, NULL, NULL, NULL } },
+    [17] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              27,
+                              { 7, -1, -1, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL, NULL },
+                              { "a3-gpio", NULL, NULL, NULL, NULL },
+                              { "a3-gpio", NULL, NULL, NULL, NULL } },
+    [18] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              30,
+                              { 7, -1, 0, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL, NULL },
+                              { "a4-gpio", NULL, NULL, NULL, NULL },
+                              { "a4-gpio", NULL, NULL, NULL, NULL } },
+    [19] = &(regmux_info_t) { PINMUX_GROUP_WAKUP,
+                              23,
+                              { 7, -1, 0, -1, -1 },
+                              { "4301c000.pinctrl-pinctrl-single", NULL, NULL, NULL, NULL },
+                              { "a5-gpio", NULL, NULL, NULL, NULL },
+                              { "a5-gpio", NULL, NULL, NULL, NULL } }
+};
+static regmux_info_t* pinmux_info_rev[MRAA_IOT2050_PINCOUNT] = { NULL };
+static regmux_info_t** pinmux_info_current;
 static mraa_gpio_context output_en_pins[MRAA_IOT2050_PINCOUNT];
 static int pull_en_pins[MRAA_IOT2050_PINCOUNT];
 
@@ -109,20 +239,6 @@ err:
     return -1;
 }
 
-static inline regmux_info_t*
-iot2050_get_regmux_by_pinmap(int pinmap)
-{
-    int i;
-
-    for(i=0; i<MRAA_IOT2050_PINCOUNT; i++) {
-        if((pinmux_info[i].mode[MUX_REGISTER_MODE_GPIO] != -1) &&
-            (pinmux_info[i].pinmap == pinmap)) {
-            return &pinmux_info[i];
-        }
-    }
-    return NULL;
-}
-
 static mraa_result_t
 iot2050_mux_debugfs(int phy_pin, const char *base_dir, const char *group, const char *function, mraa_gpio_mode_t gpio_mode)
 {
@@ -181,10 +297,9 @@ err:
 }
 
 static mraa_result_t
-iot2050_mux_mmap(int phy_pin, int mode, mraa_gpio_mode_t gpio_mode)
+iot2050_mux_mmap(const regmux_info_t* info, int mode, mraa_gpio_mode_t gpio_mode)
 {
     int8_t mux_mode;
-    regmux_info_t *info = &pinmux_info[phy_pin];
 
     syslog(LOG_ERR, "iot2050: mmap: Debugfs pinmux failed! Falling back to mmap!");
 
@@ -199,7 +314,7 @@ iot2050_mux_mmap(int phy_pin, int mode, mraa_gpio_mode_t gpio_mode)
         return MRAA_ERROR_FEATURE_NOT_SUPPORTED;
     }
 
-    syslog(LOG_DEBUG, "REGMUX[phy_pin %d] group %d index %d mode %d\n", phy_pin, info->group, info->index, mux_mode);
+    syslog(LOG_DEBUG, "REGMUX group %d index %d mode %d\n", info->group, info->index, mux_mode);
 
     platform_pinmux_select_func(pinmux_instance, info->group, info->index, mux_mode);
     /* Configure as input and output for default */
@@ -222,12 +337,18 @@ iot2050_mux_mmap(int phy_pin, int mode, mraa_gpio_mode_t gpio_mode)
 static mraa_result_t
 iot2050_mux_init_reg(int phy_pin, int mode)
 {
-    regmux_info_t *info = &pinmux_info[phy_pin];
+    regmux_info_t* info;
     int8_t mux_mode;
     mraa_result_t ret;
 
-    if((phy_pin < 0) || (phy_pin > MRAA_IOT2050_PINCOUNT) || phy_pin == 20)
+    if ((phy_pin < 0) || (phy_pin >= MRAA_IOT2050_PINCOUNT) || phy_pin == 20) {
+        return MRAA_SUCCESS;
+    }
+    if (pinmux_info_current[phy_pin] == NULL) {
         return MRAA_SUCCESS;
+    }
+    info = pinmux_info_current[phy_pin];
+
     if((mode < 0) || (mode >= MAX_MUX_REGISTER_MODE)) {
         return MRAA_ERROR_FEATURE_NOT_SUPPORTED;
     }
@@ -242,7 +363,7 @@ iot2050_mux_init_reg(int phy_pin, int mode)
 
     ret = iot2050_mux_debugfs(phy_pin, info->debugfs_path[mode], info->pmx_group[mode], info->pmx_function[mode], 0);
     if (ret != MRAA_SUCCESS)
-        return iot2050_mux_mmap(phy_pin, mode, 0);
+        return iot2050_mux_mmap(info, mode, 0);
     return ret;
 }
 
@@ -295,27 +416,79 @@ failed:
 }
 
 static mraa_result_t
-iot2050_gpio_mode_replace(mraa_gpio_context dev, mraa_gpio_mode_t mode)
+iot2050_gpio_mode_ioexpander(mraa_gpio_context dev, mraa_gpio_mode_t mode)
 {
-    regmux_info_t *info;
-    mraa_gpio_context pull_en_pin;
-    mraa_result_t ret = MRAA_SUCCESS;
+    unsigned long flags;
+    mraa_gpiod_group_t gpio_iter;
 
+    for_each_gpio_group(gpio_iter, dev)
+    {
+        int ret = mraa_gpiod_get_line_flags(gpio_iter->dev_fd, gpio_iter->gpio_lines[0], &flags);
+        if (ret != 0) {
+            syslog(LOG_ERR, "[IOT2050]: error getting line flags");
+            return MRAA_ERROR_INVALID_RESOURCE;
+        }
+
+        if (((mode == MRAA_GPIO_PULLUP) && (flags & GPIO_V2_LINE_FLAG_BIAS_PULL_UP)) ||
+            ((mode == MRAA_GPIO_PULLDOWN) && (flags & GPIO_V2_LINE_FLAG_BIAS_PULL_DOWN)) ||
+            ((mode == MRAA_GPIO_HIZ || mode == MRAA_GPIO_STRONG) && (flags & GPIO_V2_LINE_FLAG_BIAS_DISABLED))) {
+            return MRAA_SUCCESS;
+        }
+
+        switch (mode) {
+            case MRAA_GPIO_PULLUP:
+                flags |= GPIO_V2_LINE_FLAG_BIAS_PULL_UP;
+                flags &= ~GPIO_V2_LINE_FLAG_BIAS_PULL_DOWN;
+                flags &= ~GPIO_V2_LINE_FLAG_BIAS_DISABLED;
+                break;
+            case MRAA_GPIO_PULLDOWN:
+                flags |= GPIO_V2_LINE_FLAG_BIAS_PULL_DOWN;
+                flags &= ~GPIO_V2_LINE_FLAG_BIAS_PULL_UP;
+                flags &= ~GPIO_V2_LINE_FLAG_BIAS_DISABLED;
+                break;
+            case MRAA_GPIO_HIZ:
+            case MRAA_GPIO_STRONG:
+                flags |= GPIO_V2_LINE_FLAG_BIAS_DISABLED;
+                flags &= ~GPIO_V2_LINE_FLAG_BIAS_PULL_UP;
+                flags &= ~GPIO_V2_LINE_FLAG_BIAS_PULL_DOWN;
+                break;
+            default:
+                return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+        }
+
+        flags &= ~GPIO_V2_LINE_FLAG_USED;
+
+        ret = mraa_gpiod_reconfig_lines(gpio_iter->line_request_fd, flags);
+        if (ret != 0) {
+            syslog(LOG_ERR, "[IOT2050]: error reconfiguring line flags");
+            return MRAA_ERROR_INVALID_RESOURCE;
+        }
+    }
+
+    return MRAA_SUCCESS;
+}
+
+static mraa_result_t
+iot2050_gpio_mode_replace(mraa_gpio_context dev, mraa_gpio_mode_t mode)
+{
     /* Only support mode change on interface pins */
-    if(dev->phy_pin == -1) {
-        ret = MRAA_ERROR_INVALID_RESOURCE;
-        goto failed;
+    if (dev->phy_pin < 0 || dev->phy_pin >= MRAA_IOT2050_PINCOUNT) {
+        return MRAA_ERROR_INVALID_RESOURCE;
     }
     /* Handle mode changes for interface pins without pull pin */
     if (pull_en_pins[dev->phy_pin] == -1) {
-        return (mode == MRAA_GPIO_STRONG || mode == MRAA_GPIO_HIZ) ?
-            MRAA_SUCCESS : MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+        if (dev->phy_pin == 20) {
+            return MRAA_SUCCESS;
+        }
+        return iot2050_gpio_mode_ioexpander(dev, mode);
     }
-    info = iot2050_get_regmux_by_pinmap(dev->pin);
-    pull_en_pin = mraa_gpio_init_raw(pull_en_pins[dev->phy_pin]);
+
+    mraa_result_t ret = MRAA_SUCCESS;
+
+    regmux_info_t* info = pinmux_info_current[dev->phy_pin];
+    mraa_gpio_context pull_en_pin = mraa_gpio_init_raw(pull_en_pins[dev->phy_pin]);
     if(pull_en_pin == NULL) {
-        ret = MRAA_ERROR_INVALID_RESOURCE;
-        goto failed;
+        return MRAA_ERROR_INVALID_RESOURCE;
     }
     switch(mode) {
         case MRAA_GPIO_PULLUP:
@@ -326,31 +499,32 @@ iot2050_gpio_mode_replace(mraa_gpio_context dev, mraa_gpio_mode_t mode)
             if(info) {
                 ret = iot2050_mux_debugfs(dev->phy_pin, info->debugfs_path[0], info->pmx_group[0], info->pmx_function[0], mode);
                 if (ret != MRAA_SUCCESS)
-                    ret = iot2050_mux_mmap(dev->phy_pin, 0, mode);
-
+                    ret = iot2050_mux_mmap(info, 0, mode);
             }
             break;
         case MRAA_GPIO_PULLDOWN:
-            if(mraa_gpio_dir(pull_en_pin, MRAA_GPIO_OUT_LOW) != MRAA_SUCCESS) {
+            if (mraa_gpio_dir(pull_en_pin, MRAA_GPIO_OUT_LOW) != MRAA_SUCCESS) {
                 ret = MRAA_ERROR_INVALID_RESOURCE;
                 goto failed;
             }
-            if(info) {
-                ret = iot2050_mux_debugfs(dev->phy_pin, info->debugfs_path[0], info->pmx_group[0], info->pmx_function[0], mode);
+            if (info) {
+                ret = iot2050_mux_debugfs(dev->phy_pin, info->debugfs_path[0], info->pmx_group[0],
+                                          info->pmx_function[0], mode);
                 if (ret != MRAA_SUCCESS)
-                    ret = iot2050_mux_mmap(dev->phy_pin, 0, mode);
+                    ret = iot2050_mux_mmap(info, 0, mode);
             }
             break;
         case MRAA_GPIO_HIZ:
         case MRAA_GPIO_STRONG:
-            if(mraa_gpio_dir(pull_en_pin, MRAA_GPIO_IN) != MRAA_SUCCESS) {
+            if (mraa_gpio_dir(pull_en_pin, MRAA_GPIO_IN) != MRAA_SUCCESS) {
                 ret = MRAA_ERROR_INVALID_RESOURCE;
                 goto failed;
             }
-            if(info) {
-                ret = iot2050_mux_debugfs(dev->phy_pin, info->debugfs_path[0], info->pmx_group[0], info->pmx_function[0], mode);
+            if (info) {
+                ret = iot2050_mux_debugfs(dev->phy_pin, info->debugfs_path[0], info->pmx_group[0],
+                                          info->pmx_function[0], mode);
                 if (ret != MRAA_SUCCESS)
-                    ret = iot2050_mux_mmap(dev->phy_pin, 0, mode);
+                    ret = iot2050_mux_mmap(info, 0, mode);
             }
             break;
         case MRAA_GPIOD_ACTIVE_LOW:
@@ -364,7 +538,7 @@ iot2050_gpio_mode_replace(mraa_gpio_context dev, mraa_gpio_mode_t mode)
     return ret;
 failed:
     syslog(LOG_ERR, "iot2050: Error setting gpio mode");
-    if(pull_en_pin) {
+    if (pull_en_pin) {
         mraa_gpio_close(pull_en_pin);
     }
     return ret;
@@ -399,22 +573,18 @@ iot2050_gpio_init_internal_replace(mraa_gpio_context dev, int pin)
     return MRAA_SUCCESS;
 }
 static inline void
-iot2050_setup_pins(mraa_board_t *board, int pin_index, char *pin_name, mraa_pincapabilities_t cap, regmux_info_t mux_info)
+iot2050_setup_pins(mraa_board_t* board, int pin_index, char* pin_name, mraa_pincapabilities_t cap)
 {
     strncpy(board->pins[pin_index].name, pin_name, MRAA_PIN_NAME_SIZE);
     board->pins[pin_index].capabilities = cap;
-    pinmux_info[pin_index] = mux_info;
 }
 
 static inline void
-iot2050_pin_add_gpio(mraa_board_t *board, int pin_index, int chip, int line,
-                        int output_en_pin, int pull_en_pin, mraa_mux_t *pin_mux, int num_pinmux)
+iot2050_pin_add_gpio(mraa_board_t* board, int pin_index, int chip, int line, int output_en_pin, int pull_en_pin, mraa_mux_t* pin_mux, int num_pinmux)
 {
     int i;
 
-    mraa_pininfo_t *pininfo = &(board->pins[pin_index]);
-    /*for sysfs*/
-    pininfo->gpio.pinmap = pinmux_info[pin_index].pinmap;
+    mraa_pininfo_t* pininfo = &(board->pins[pin_index]);
     pininfo->gpio.mux_total = 0;
     /*for gpio chardev*/
     pininfo->gpio.gpio_chip = chip;
@@ -559,25 +729,26 @@ mraa_board_t*
 mraa_siemens_iot2050()
 {
     mraa_board_t* b = (mraa_board_t*) calloc(1, sizeof(mraa_board_t));
-    mraa_mux_t mux_info[6];
+    mraa_mux_t mux_info[6] = { 0 };
     int pin_index = 0;
-    unsigned main_gpio0_chip, main_gpio0_base;
+    unsigned main_gpio0_chip;
     unsigned wkup_gpio0_chip, wkup_gpio0_base;
     unsigned d4200_gpio_chip, d4200_gpio_base;
     unsigned d4201_gpio_chip, d4201_gpio_base;
     unsigned d4202_gpio_chip, d4202_gpio_base;
     unsigned line_offset;
     int res;
+    int is_rev = 0;
 
-    if (mraa_find_gpio_line_by_name("main_gpio0-base", &main_gpio0_chip, &line_offset) < 0 || line_offset != 0) {
-        goto error;
+    if (mraa_file_contains("/proc/device-tree/model", "SIMATIC IOT2050 Advanced PG2 Rev") ||
+        mraa_file_contains("/proc/device-tree/model", "SIMATIC IOT2050 Advanced M2 Rev")) {
+        is_rev = 1;
     }
-    res = mraa_get_chip_base_by_number(main_gpio0_chip);
-    if (res < 0) {
+    pinmux_info_current = is_rev ? pinmux_info_rev : pinmux_info;
+
+    if (mraa_find_gpio_line_by_name("main_gpio0-base", &main_gpio0_chip, &line_offset) < 0 || line_offset != 0) {
         goto error;
     }
-    main_gpio0_base = res;
-
     if (mraa_find_gpio_line_by_name("wkup_gpio0-base", &wkup_gpio0_chip, &line_offset) < 0 || line_offset != 0) {
         goto error;
     }
@@ -587,8 +758,9 @@ mraa_siemens_iot2050()
     }
     wkup_gpio0_base = res;
 
-    /* A0-pull is line 0 in D4200 */
-    if (mraa_find_gpio_line_by_name("A0-pull", &d4200_gpio_chip, &line_offset) < 0 || line_offset != 0) {
+    /* A0/A0-pull is line 0 in D4200 */
+    if (mraa_find_gpio_line_by_name(is_rev ? "A0" : "A0-pull", &d4200_gpio_chip, &line_offset) < 0 ||
+        line_offset != 0) {
         goto error;
     }
     res = mraa_get_chip_base_by_number(d4200_gpio_chip);
@@ -607,8 +779,9 @@ mraa_siemens_iot2050()
     }
     d4201_gpio_base = res;
 
-    /* IO0-pull is line 0 in D4202 */
-    if (mraa_find_gpio_line_by_name("IO0-pull", &d4202_gpio_chip, &line_offset) < 0 || line_offset != 0) {
+    /* IO0/IO0-pull is line 0 in D4202 */
+    if (mraa_find_gpio_line_by_name(is_rev ? "IO0" : "IO0-pull", &d4202_gpio_chip, &line_offset) < 0 ||
+        line_offset != 0) {
         goto error;
     }
     res = mraa_get_chip_base_by_number(d4202_gpio_chip);
@@ -623,7 +796,7 @@ mraa_siemens_iot2050()
     memset(output_en_pins, 0, sizeof(mraa_gpio_context) * MRAA_IOT2050_PINCOUNT);
     b->platform_name = PLATFORM_NAME;
     b->phy_pin_count = MRAA_IOT2050_PINCOUNT;
-    b->chardev_capable = 1;
+    b->chardev_ver = 2;
     b->adc_raw = 12;
     b->adc_supported = 12;
     b->pwm_default_period = 1000; /*us*/
@@ -644,1201 +817,552 @@ mraa_siemens_iot2050()
         goto error;
     }
 
-    /* IO */
-    iot2050_setup_pins(b, pin_index, "IO0",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            1}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            17,
-                            wkup_gpio0_base+29,
-                            {
-                                7,  /*GPIO*/
-                                4,  /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d0-gpio",
-                                "d0-uart0-rxd",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d0-gpio",
-                                "d0-uart0-rxd",
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 29, d4201_gpio_base+0, d4202_gpio_base+0, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+0;
+    iot2050_setup_pins(b, pin_index, "IO0", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 0, 1 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 0;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 0, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 29, d4201_gpio_base + 0,
+                             d4202_gpio_base + 0, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 0;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    mux_info[1].pin = d4202_gpio_base+0;
+    mux_info[1].pin = d4202_gpio_base + 0;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_uart(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO1",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            1}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            18,
-                            wkup_gpio0_base+30,
-                            {
-                                7,  /*GPIO*/
-                                4,  /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d1-gpio",
-                                "d1-uart0-txd",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d1-gpio",
-                                "d1-uart0-txd",
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 30, d4201_gpio_base+1, d4202_gpio_base+1, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+1;
+    iot2050_setup_pins(b, pin_index, "IO1", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 0, 1 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 1;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 1, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 30, d4201_gpio_base + 1,
+                             d4202_gpio_base + 1, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 1;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+1;
+    mux_info[1].pin = d4202_gpio_base + 1;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_uart(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO2",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            1}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            19,
-                            wkup_gpio0_base+31,
-                            {
-                                7,  /*GPIO*/
-                                4,  /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d2-gpio",
-                                "d2-uart0-ctsn",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d2-gpio",
-                                "d2-uart0-ctsn",
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 31, d4201_gpio_base+2, d4202_gpio_base+2, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+2;
+    iot2050_setup_pins(b, pin_index, "IO2", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 0, 1 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 2;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 2, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 31, d4201_gpio_base + 2,
+                             d4202_gpio_base + 2, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 2;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    mux_info[1].pin = d4202_gpio_base+2;
+    mux_info[1].pin = d4202_gpio_base + 2;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_uart(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO3",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            1}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            21,
-                            wkup_gpio0_base+33,
-                            {
-                                7,  /*GPIO*/
-                                4,  /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d3-gpio",
-                                "d3-uart0-rtsn",
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "d3-gpio",
-                                "d3-uart0-rtsn",
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 33, d4201_gpio_base+3, d4202_gpio_base+3, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+3;
+    iot2050_setup_pins(b, pin_index, "IO3", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 0, 1 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 3;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 3, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 33, d4201_gpio_base + 3,
+                             d4202_gpio_base + 3, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 3;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+3;
+    mux_info[1].pin = d4202_gpio_base + 3;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_uart(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO4",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            1,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_MAIN,
-                            33,
-                            main_gpio0_base+33,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                5   /*PWM*/
-                            },
-                            {
-                                "11c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "11c000.pinctrl-pinctrl-single"
-                            },
-                            {
-                                "d4-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d4-ehrpwm0-a"
-                            },
-                            {
-                                "d4-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d4-ehrpwm0-a"
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 33, d4201_gpio_base+4, d4202_gpio_base+4, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+4;
+    iot2050_setup_pins(b, pin_index, "IO4", (mraa_pincapabilities_t) { 1, 1, 1, 0, 0, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 4;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 4, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 33, d4201_gpio_base + 4,
+                             d4202_gpio_base + 4, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 4;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+4;
+    mux_info[1].pin = d4202_gpio_base + 4;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_pwm(b, pin_index, 0, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO5",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            1,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_MAIN,
-                            35,
-                            main_gpio0_base+35,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                5   /*PWM*/
-                            },
-                            {
-                                "11c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "11c000.pinctrl-pinctrl-single"
-                            },
-                            {
-                                "d5-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d5-ehrpwm1-a"
-                            },
-                            {
-                                "d5-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d5-ehrpwm1-a"
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 35, d4201_gpio_base+5, d4202_gpio_base+5, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+5;
+    iot2050_setup_pins(b, pin_index, "IO5", (mraa_pincapabilities_t) { 1, 1, 1, 0, 0, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 5;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 5, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 35, d4201_gpio_base + 5,
+                             d4202_gpio_base + 5, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 5;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+5;
+    mux_info[1].pin = d4202_gpio_base + 5;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_pwm(b, pin_index, 1, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO6",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            1,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_MAIN,
-                            38,
-                            main_gpio0_base+38,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                5   /*PWM*/
-                            },
-                            {
-                                "11c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "11c000.pinctrl-pinctrl-single"
-                            },
-                            {
-                                "d6-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d6-ehrpwm2-a"
-                            },
-                            {
-                                "d6-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d6-ehrpwm2-a"
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 38, d4201_gpio_base+6, d4202_gpio_base+6, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+6;
+    iot2050_setup_pins(b, pin_index, "IO6", (mraa_pincapabilities_t) { 1, 1, 1, 0, 0, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 6;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 6, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 38, d4201_gpio_base + 6,
+                             d4202_gpio_base + 6, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 6;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+6;
+    mux_info[1].pin = d4202_gpio_base + 6;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_pwm(b, pin_index, 2, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO7",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            1,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_MAIN,
-                            43,
-                            main_gpio0_base+43,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                5   /*PWM*/
-                            },
-                            {
-                                "11c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "11c000.pinctrl-pinctrl-single"
-                            },
-                            {
-                                "d7-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d7-ehrpwm3-a"
-                            },
-                            {
-                                "d7-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d7-ehrpwm3-a"
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 43, d4201_gpio_base+7, d4202_gpio_base+7, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+7;
+    iot2050_setup_pins(b, pin_index, "IO7", (mraa_pincapabilities_t) { 1, 1, 1, 0, 0, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 7;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 7, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 43, d4201_gpio_base + 7,
+                             d4202_gpio_base + 7, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 7;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+7;
+    mux_info[1].pin = d4202_gpio_base + 7;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_pwm(b, pin_index, 3, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO8",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            1,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_MAIN,
-                            48,
-                            main_gpio0_base+48,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                5   /*PWM*/
-                            },
-                            {
-                                "11c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "11c000.pinctrl-pinctrl-single"
-                            },
-                            {
-                                "d8-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d8-ehrpwm4-a"
-                            },
-                            {
-                                "d8-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d8-ehrpwm4-a"
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 48, d4201_gpio_base+8, d4202_gpio_base+8, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+8;
+    iot2050_setup_pins(b, pin_index, "IO8", (mraa_pincapabilities_t) { 1, 1, 1, 0, 0, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 8;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 8, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 48, d4201_gpio_base + 8,
+                             d4202_gpio_base + 8, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 8;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+8;
+    mux_info[1].pin = d4202_gpio_base + 8;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_pwm(b, pin_index, 4, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO9",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            1,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_MAIN,
-                            51,
-                            main_gpio0_base+51,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                5   /*PWM*/
-                            },
-                            {
-                                "11c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "11c000.pinctrl-pinctrl-single"
-                            },
-                            {
-                                "d9-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d9-ehrpwm5-a"
-                            },
-                            {
-                                "d9-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                "d9-ehrpwm5-a"
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 51, d4201_gpio_base+9, d4202_gpio_base+9, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+9;
+    iot2050_setup_pins(b, pin_index, "IO9", (mraa_pincapabilities_t) { 1, 1, 1, 0, 0, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 9;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 9, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, main_gpio0_chip, 51, d4201_gpio_base + 9,
+                             d4202_gpio_base + 9, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 9;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+9;
+    mux_info[1].pin = d4202_gpio_base + 9;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_pwm(b, pin_index, 5, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO10",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            1,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            39,
-                            wkup_gpio0_base+51,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                0, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL
-                            },
-                            {
-                                "d10-gpio",
-                                NULL,
-                                NULL,
-                                "d10-spi0-cs0",
-                                NULL
-                            },
-                            {
-                                "d10-gpio",
-                                NULL,
-                                NULL,
-                                "d10-spi0-cs0",
-                                NULL
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 51, d4201_gpio_base+10, d4202_gpio_base+10, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+10;
+    iot2050_setup_pins(b, pin_index, "IO10", (mraa_pincapabilities_t) { 1, 1, 0, 0, 1, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 10;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 10, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 51, d4201_gpio_base + 10,
+                             d4202_gpio_base + 10, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 10;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+10;
+    mux_info[1].pin = d4202_gpio_base + 10;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_spi(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO11",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            1,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            37,
-                            wkup_gpio0_base+49,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                0, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL
-                            },
-                            {
-                                "d11-gpio",
-                                NULL,
-                                NULL,
-                                "d11-spi0-d0",
-                                NULL
-                            },
-                            {
-                                "d11-gpio",
-                                NULL,
-                                NULL,
-                                "d11-spi0-d0",
-                                NULL
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 49, d4201_gpio_base+11, d4202_gpio_base+11, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+11;
+    iot2050_setup_pins(b, pin_index, "IO11", (mraa_pincapabilities_t) { 1, 1, 0, 0, 1, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 11;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 11, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 49, d4201_gpio_base + 11,
+                             d4202_gpio_base + 11, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 11;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+11;
+    mux_info[1].pin = d4202_gpio_base + 11;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_spi(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO12",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            1,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            38,
-                            wkup_gpio0_base+50,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                0, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL
-                            },
-                            {
-                                "d12-gpio",
-                                NULL,
-                                NULL,
-                                "d12-spi0-d1",
-                                NULL
-                            },
-                            {
-                                "d12-gpio",
-                                NULL,
-                                NULL,
-                                "d12-spi0-d1",
-                                NULL
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 50, d4201_gpio_base+12, d4202_gpio_base+12, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+12;
+    iot2050_setup_pins(b, pin_index, "IO12", (mraa_pincapabilities_t) { 1, 1, 0, 0, 1, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 12;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 12, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 50, d4201_gpio_base + 12,
+                             d4202_gpio_base + 12, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 12;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    mux_info[1].pin = d4202_gpio_base+12;
+    mux_info[1].pin = d4202_gpio_base + 12;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_spi(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "IO13",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            1,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            36,
-                            wkup_gpio0_base+48,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                0, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL
-                            },
-                            {
-                                "d13-gpio",
-                                NULL,
-                                NULL,
-                                "d13-spi0-clk",
-                                NULL
-                            },
-                            {
-                                "d13-gpio",
-                                NULL,
-                                NULL,
-                                "d13-spi0-clk",
-                                NULL
-                            }
-                        });
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 48, d4201_gpio_base+13, d4202_gpio_base+13, NULL, 0);
-    mux_info[0].pin = d4201_gpio_base+13;
+    iot2050_setup_pins(b, pin_index, "IO13", (mraa_pincapabilities_t) { 1, 1, 0, 0, 1, 0, 0, 0 });
+    if (is_rev) {
+        mux_info[0].pin = d4201_gpio_base + 13;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, d4202_gpio_chip, 13, -1, -1, mux_info, 1);
+    } else {
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 48, d4201_gpio_base + 13,
+                             d4202_gpio_base + 13, NULL, 0);
+    }
+    mux_info[0].pin = d4201_gpio_base + 13;
     mux_info[0].pincmd = PINCMD_SET_DIRECTION;
     mux_info[0].value = MRAA_GPIO_OUT_HIGH;
-    mux_info[1].pin = d4202_gpio_base+13;
+    mux_info[1].pin = d4202_gpio_base + 13;
     mux_info[1].pincmd = PINCMD_SET_DIRECTION;
     mux_info[1].value = MRAA_GPIO_IN;
     iot2050_pin_add_spi(b, pin_index, 0, mux_info, 2);
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "A0",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            1,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            33,
-                            wkup_gpio0_base+45,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a0-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a0-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    mux_info[0].pin = d4200_gpio_base+8;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 45, wkup_gpio0_base+38, d4200_gpio_base+0, mux_info, 1);
-    // D/A switch
-    mux_info[0].pin = d4200_gpio_base+8;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    // muxed GPIO as input
-    mux_info[1].pin = wkup_gpio0_base+45;
-    mux_info[1].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[1].value = MRAA_GPIO_IN;
-    // output enable as input
-    mux_info[2].pin = wkup_gpio0_base+38;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_OUT_LOW;
-    // pull enable as input
-    mux_info[3].pin = d4200_gpio_base+0;
-    mux_info[3].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[3].value = MRAA_GPIO_IN;
-    iot2050_pin_add_aio(b, pin_index, 0, mux_info, 4);
+    iot2050_setup_pins(b, pin_index, "A0", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 1, 0 });
+    if (is_rev) {
+        /* IO14_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 8;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        iot2050_pin_add_gpio(b, pin_index, d4200_gpio_chip, 0, -1, -1, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 8;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = d4200_gpio_base + 0;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 0, mux_info, 2);
+    } else {
+        mux_info[0].pin = wkup_gpio0_base + 45;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 45, wkup_gpio0_base + 38,
+                             d4200_gpio_base + 0, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 8;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 45;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        mux_info[2].pin = wkup_gpio0_base + 38;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_OUT_LOW;
+        mux_info[3].pin = d4200_gpio_base + 0;
+        mux_info[3].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[3].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 0, mux_info, 4);
+    }
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "A1",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            1,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            32,
-                            wkup_gpio0_base+44,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a1-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a1-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    mux_info[0].pin = d4200_gpio_base+9;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 44, wkup_gpio0_base+37, d4200_gpio_base+1, mux_info, 1);
-    // D/A switch
-    mux_info[0].pin = d4200_gpio_base+9;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    // muxed GPIO as input
-    mux_info[1].pin = wkup_gpio0_base+44;
-    mux_info[1].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[1].value = MRAA_GPIO_IN;
-    // output enable as input
-    mux_info[2].pin = wkup_gpio0_base+37;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_OUT_LOW;
-    // pull enable as input
-    mux_info[3].pin = d4200_gpio_base+1;
-    mux_info[3].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[3].value = MRAA_GPIO_IN;
-    iot2050_pin_add_aio(b, pin_index, 1, mux_info, 4);
+    iot2050_setup_pins(b, pin_index, "A1", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 1, 0 });
+    if (is_rev) {
+        /* IO15_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 9;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        iot2050_pin_add_gpio(b, pin_index, d4200_gpio_chip, 1, -1, -1, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 9;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = d4200_gpio_base + 1;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 1, mux_info, 2);
+    } else {
+        mux_info[0].pin = wkup_gpio0_base + 44;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 44, wkup_gpio0_base + 37, -1, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 9;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 44;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        mux_info[2].pin = wkup_gpio0_base + 37;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_OUT_LOW;
+        mux_info[3].pin = d4200_gpio_base + 1;
+        mux_info[3].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[3].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 1, mux_info, 4);
+    }
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "A2",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            1,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            31,
-                            wkup_gpio0_base+43,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a2-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a2-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    mux_info[0].pin = d4200_gpio_base+10;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 43, wkup_gpio0_base+36, d4200_gpio_base+2, mux_info, 1);
-    // D/A switch
-    mux_info[0].pin = d4200_gpio_base+10;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    // muxed GPIO as input
-    mux_info[1].pin = wkup_gpio0_base+43;
-    mux_info[1].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[1].value = MRAA_GPIO_IN;
-    // output enable as input
-    mux_info[2].pin = wkup_gpio0_base+36;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_OUT_LOW;
-    // pull enable as input
-    mux_info[3].pin = d4200_gpio_base+2;
-    mux_info[3].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[3].value = MRAA_GPIO_IN;
-    iot2050_pin_add_aio(b, pin_index, 2, mux_info, 4);
+    iot2050_setup_pins(b, pin_index, "A2", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 1, 0 });
+    if (is_rev) {
+        /* IO16_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 10;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        iot2050_pin_add_gpio(b, pin_index, d4200_gpio_chip, 2, -1, -1, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 10;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = d4200_gpio_base + 2;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 2, mux_info, 2);
+    } else {
+        mux_info[0].pin = wkup_gpio0_base + 43;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 43, wkup_gpio0_base + 36,
+                             d4200_gpio_base + 2, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 10;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 43;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        mux_info[2].pin = wkup_gpio0_base + 36;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_OUT_LOW;
+        mux_info[3].pin = d4200_gpio_base + 2;
+        mux_info[3].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[3].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 2, mux_info, 4);
+    }
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "A3",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            1,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            27,
-                            wkup_gpio0_base+39,
-                            {
-                                7,  /*GPIO*/
-                                -1, /*UART*/
-                                -1, /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a3-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a3-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    mux_info[0].pin = d4200_gpio_base+11;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 39, wkup_gpio0_base+34, d4200_gpio_base+3, mux_info, 1);
-    // D/A switch
-    mux_info[0].pin = d4200_gpio_base+11;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    // muxed GPIO as input
-    mux_info[1].pin = wkup_gpio0_base+39;
-    mux_info[1].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[1].value = MRAA_GPIO_IN;
-    // output enable as input
-    mux_info[2].pin = wkup_gpio0_base+34;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_OUT_LOW;
-    // pull enable as input
-    mux_info[3].pin = d4200_gpio_base+3;
-    mux_info[3].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[3].value = MRAA_GPIO_IN;
-    iot2050_pin_add_aio(b, pin_index, 3, mux_info, 4);
+    iot2050_setup_pins(b, pin_index, "A3", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 1, 0 });
+    if (is_rev) {
+        /* IO17_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 11;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        iot2050_pin_add_gpio(b, pin_index, d4200_gpio_chip, 3, -1, -1, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 11;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = d4200_gpio_base + 3;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 3, mux_info, 2);
+    } else {
+        mux_info[0].pin = wkup_gpio0_base + 39;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 39, wkup_gpio0_base + 34,
+                             d4200_gpio_base + 3, mux_info, 1);
+        mux_info[0].pin = d4200_gpio_base + 11;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 39;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        mux_info[2].pin = wkup_gpio0_base + 34;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_OUT_LOW;
+        mux_info[3].pin = d4200_gpio_base + 3;
+        mux_info[3].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[3].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 3, mux_info, 4);
+    }
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "A4",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            1,  /*i2c*/
-                            1,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            30,
-                            wkup_gpio0_base+42,
-                            {
-                                7, /*GPIO*/
-                                -1, /*UART*/
-                                0,  /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a4-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a4-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    mux_info[0].pin = d4200_gpio_base+12;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    mux_info[1].pin = wkup_gpio0_base+21;
-    mux_info[1].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[1].value = MRAA_GPIO_OUT_HIGH;
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 42, wkup_gpio0_base+41, d4200_gpio_base+4, mux_info, 2);
-    // A/D/I switch
-    mux_info[0].pin = d4200_gpio_base+12;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    mux_info[1].pin = wkup_gpio0_base+21;
-    mux_info[1].pincmd = PINCMD_SET_OUT_VALUE;
-    mux_info[1].value = 0;
-    // pull enable as input
-    mux_info[2].pin = d4200_gpio_base+4;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_IN;
-    iot2050_pin_add_i2c(b, pin_index, 0, mux_info, 3);
-    // A/D/I switch
-    mux_info[0].pin = d4200_gpio_base+12;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    mux_info[1].pin = wkup_gpio0_base+21;
-    mux_info[1].pincmd = PINCMD_SET_OUT_VALUE;
-    mux_info[1].value = 1;
-    // muxed GPIO as input
-    mux_info[2].pin = wkup_gpio0_base+42;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_IN;
-    // output enable as input
-    mux_info[3].pin = wkup_gpio0_base+41;
-    mux_info[3].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[3].value = MRAA_GPIO_OUT_LOW;
-    // pull enable as input
-    mux_info[4].pin = d4200_gpio_base+4;
-    mux_info[4].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[4].value = MRAA_GPIO_IN;
-    iot2050_pin_add_aio(b, pin_index, 4, mux_info, 5);
+    iot2050_setup_pins(b, pin_index, "A4", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 1, 1, 0 });
+    if (is_rev) {
+        /* IO18_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 12;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        /* AMUX_IN */
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_HIGH;
+        iot2050_pin_add_gpio(b, pin_index, d4200_gpio_chip, 4, -1, -1, mux_info, 2);
+        /* IO18_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 12;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        /* AMUX_IN */
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_HIGH;
+        /* IO18 GPIO should be input */
+        mux_info[2].pin = d4200_gpio_base + 4;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 4, mux_info, 3);
+        /* IO18_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 12;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        /* AMUX_IN */
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_i2c(b, pin_index, 0, mux_info, 2);
+    } else {
+        mux_info[0].pin = d4200_gpio_base + 12;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_HIGH;
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 42, wkup_gpio0_base + 41,
+                             d4200_gpio_base + 4, mux_info, 2);
+        mux_info[0].pin = d4200_gpio_base + 12;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 42;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        mux_info[2].pin = wkup_gpio0_base + 41;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_OUT_LOW;
+        mux_info[3].pin = d4200_gpio_base + 4;
+        mux_info[3].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[3].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 4, mux_info, 5);
+        mux_info[0].pin = d4200_gpio_base + 12;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_OUT_VALUE;
+        mux_info[1].value = 0;
+        mux_info[2].pin = d4200_gpio_base + 4;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_IN;
+        iot2050_pin_add_i2c(b, pin_index, 0, mux_info, 3);
+    }
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "A5",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            1,  /*i2c*/
-                            1,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            PINMUX_GROUP_WAKUP,
-                            23,
-                            wkup_gpio0_base+35,
-                            {
-                                7, /*GPIO*/
-                                -1, /*UART*/
-                                0,  /*I2C*/
-                                -1, /*SPI*/
-                                -1  /*PWM*/
-                            },
-                            {
-                                "4301c000.pinctrl-pinctrl-single",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a5-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            },
-                            {
-                                "a5-gpio",
-                                NULL,
-                                NULL,
-                                NULL,
-                                NULL
-                            }
-                        });
-    mux_info[0].pin = d4200_gpio_base+13;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_OUT_LOW;
-    mux_info[1].pin = wkup_gpio0_base+21;
-    mux_info[1].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[1].value = MRAA_GPIO_OUT_HIGH;
-    iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 35, d4201_gpio_base+14, d4200_gpio_base+5, mux_info, 2);
-    // A/D/I switch
-    mux_info[0].pin = d4200_gpio_base+13;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    mux_info[1].pin = wkup_gpio0_base+21;
-    mux_info[1].pincmd = PINCMD_SET_OUT_VALUE;
-    mux_info[1].value = 0;
-    // pull enable as input
-    mux_info[2].pin = d4200_gpio_base+5;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_IN;
-    iot2050_pin_add_i2c(b, pin_index, 0, mux_info, 3);
-    // A/D/I switch
-    mux_info[0].pin = d4200_gpio_base+13;
-    mux_info[0].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[0].value = MRAA_GPIO_IN;
-    mux_info[1].pin = wkup_gpio0_base+21;
-    mux_info[1].pincmd = PINCMD_SET_OUT_VALUE;
-    mux_info[1].value = 1;
-    // muxed GPIO as input
-    mux_info[2].pin = wkup_gpio0_base+35;
-    mux_info[2].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[2].value = MRAA_GPIO_IN;
-    // output enable as input
-    mux_info[3].pin = d4201_gpio_base+14;
-    mux_info[3].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[3].value = MRAA_GPIO_OUT_LOW;
-    // pull enable as input
-    mux_info[4].pin = d4200_gpio_base+5;
-    mux_info[4].pincmd = PINCMD_SET_DIRECTION;
-    mux_info[4].value = MRAA_GPIO_IN;
-    iot2050_pin_add_aio(b, pin_index, 5, mux_info, 5);
+    iot2050_setup_pins(b, pin_index, "A5", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 1, 1, 0 });
+    if (is_rev) {
+        /* IO19_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 13;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        /* AMUX_IN */
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_HIGH;
+        iot2050_pin_add_gpio(b, pin_index, d4200_gpio_chip, 5, -1, -1, mux_info, 2);
+        /* IO19_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 13;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        /* AMUX_IN */
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_HIGH;
+        /* IO19 GPIO should be input */
+        mux_info[2].pin = d4200_gpio_base + 5;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 5, mux_info, 3);
+        /* IO19_EN# PIN */
+        mux_info[0].pin = d4200_gpio_base + 13;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        /* AMUX_IN */
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_LOW;
+        iot2050_pin_add_i2c(b, pin_index, 0, mux_info, 2);
+    } else {
+        mux_info[0].pin = d4200_gpio_base + 13;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_OUT_LOW;
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_OUT_HIGH;
+        iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 35, d4201_gpio_base + 14,
+                             d4200_gpio_base + 5, mux_info, 2);
+        mux_info[0].pin = d4200_gpio_base + 13;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 35;
+        mux_info[1].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[1].value = MRAA_GPIO_IN;
+        mux_info[2].pin = d4201_gpio_base + 14;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_OUT_LOW;
+        mux_info[3].pin = d4200_gpio_base + 5;
+        mux_info[3].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[3].value = MRAA_GPIO_IN;
+        iot2050_pin_add_aio(b, pin_index, 5, mux_info, 5);
+        mux_info[0].pin = d4200_gpio_base + 13;
+        mux_info[0].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[0].value = MRAA_GPIO_IN;
+        mux_info[1].pin = wkup_gpio0_base + 21;
+        mux_info[1].pincmd = PINCMD_SET_OUT_VALUE;
+        mux_info[1].value = 0;
+        mux_info[2].pin = d4200_gpio_base + 5;
+        mux_info[2].pincmd = PINCMD_SET_DIRECTION;
+        mux_info[2].value = MRAA_GPIO_IN;
+        iot2050_pin_add_i2c(b, pin_index, 0, mux_info, 3);
+    }
     pin_index++;
 
-    iot2050_setup_pins(b, pin_index, "USER",
-                        (mraa_pincapabilities_t) {
-                            1,  /*valid*/
-                            1,  /*gpio*/
-                            0,  /*pwm*/
-                            0,  /*fast gpio*/
-                            0,  /*spi*/
-                            0,  /*i2c*/
-                            0,  /*aio*/
-                            0}, /*uart*/
-                        (regmux_info_t) {
-                            -1,
-                            -1,
-                            wkup_gpio0_base+25,
-                            {0},
-                            {0},
-                            {0},
-                            {0}
-                        });
+    iot2050_setup_pins(b, pin_index, "USER", (mraa_pincapabilities_t) { 1, 1, 0, 0, 0, 0, 0, 0 });
     iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 25, -1, -1, NULL, 0);
     pin_index++;
 
@@ -1879,6 +1403,8 @@ mraa_siemens_iot2050_sm()
     unsigned line_offset;
     mraa_board_t* b = (mraa_board_t*) calloc(1, sizeof(mraa_board_t));
 
+    pinmux_info_current = pinmux_info_rev;
+
     if (NULL == b) {
         goto error_board;
     }
@@ -1905,24 +1431,8 @@ mraa_siemens_iot2050_sm()
 
     /* USER BUTTON */
     iot2050_setup_pins(b, pin_index, "USER",
-                        (mraa_pincapabilities_t) {
-                            .valid = 1,
-                            .gpio = 1,
-                            .pwm = 0,
-                            .fast_gpio = 0,
-                            .spi = 0,
-                            .i2c = 0,
-                            .aio = 0,
-                            .uart = 0},
-                        (regmux_info_t) {
-                            .group = -1,
-                            .index = -1,
-                            .pinmap = wkup_gpio0_base+25,
-                            .mode = {0},
-                            .debugfs_path = {0},
-                            .pmx_function = {0},
-                            .pmx_group = {0},
-                        });
+                       (mraa_pincapabilities_t) {
+                       .valid = 1, .gpio = 1, .pwm = 0, .fast_gpio = 0, .spi = 0, .i2c = 0, .aio = 0, .uart = 0 });
     iot2050_pin_add_gpio(b, pin_index, wkup_gpio0_chip, 25, -1, -1, NULL, 0);
 
     /* LED */
